### ISSUE TITLE
[detailed] Unified EMO + EMS Design

### ISSUE THREAD

**Issue by TroyGarden** [pytorch/torchrec#4100]
# Unified EMO \+ EMS Design

## 1\. Motivation

Embedding Memory Offloading (EMO) and Embedding Memory Stashing (EMS) both address HBM pressure from embedding tables, but at different levels:

| Feature | Mechanism | HBM Usage |
| :---- | :---- | :---- |
| EMO | fused\_uvm\_caching — full table in UVM (host DDR), hot rows cached in HBM via lxu\_cache\_weights | Cache lives in HBM permanently |
| Embedding Stashing | After forward, async copy weights\_dev GPU→CPU, free HBM via resize\_(0), restore before backward | HBM freed between forward and backward |

**Problem**: These two features are currently incompatible. For fused\_uvm\_caching tables, weights\_dev is an empty tensor — the actual HBM consumer is lxu\_cache\_weights. Stashing weights\_dev is a no-op.

**Goal**: Time-share HBM between the embedding cache and dense compute. Free the cache during dense forward/backward, restore it before sparse backward.

## 2\. Background: FBGEMM TBE Cache Architecture

For EmbeddingLocation.MANAGED\_CACHING tables:

| Tensor | Location | Purpose |
| :---- | :---- | :---- |
| weights\_dev | GPU HBM | Empty — unused for MANAGED\_CACHING |
| weights\_uvm | Host DDR (UVM) | Full embedding table, authoritative copy |
| lxu\_cache\_weights | GPU HBM | Set-associative cache (32-way) of hot rows |
| lxu\_cache\_state | GPU HBM | Tracks which embedding row occupies each cache slot |
| lxu\_state | GPU HBM | LRU timestamps or LFU access counters |

### 2.1. Cache Lifecycle in Standard EMO

```
prefetch(batch_i):
  linearize_cache_indices()      — map per-table indices to flat cache index space
  lru_cache_populate()           — evict LRU victims (write dirty → weights_uvm),
                                   load needed rows from weights_uvm → lxu_cache_weights
  lxu_cache_lookup()             — produce lxu_cache_locations: index → cache slot

forward(batch_i):
  read from lxu_cache_weights[lxu_cache_locations]

backward(batch_i):
  TBE fused backward+optimizer   — updates lxu_cache_weights in-place (rows become DIRTY)
```

### 2.2. Dirty Lines

Cache entries modified by the TBE fused backward+optimizer but not yet written back to weights\_uvm. They are only flushed on eviction (during a future prefetch) or via explicit flush().

### 2.3. Fused Backward+Optimizer

There is no separate optimizer.step() for embeddings. The TBE backward kernel fuses gradient computation and optimizer update (SGD/Adam/etc.) into one kernel that writes directly to lxu\_cache\_weights.

## 3\. Proposed Design: Unified EMO \+ Stash Pipeline

The core idea: move the EMO prefetch into the stash-restore step. Instead of prefetching before forward, prefetch at restore time (before backward). The cache is freed during dense compute and only holds rows for the current \+ next batch.

### 3.1. Pipeline Flow

```
                    Iteration i

forward(batch_i)
  └─ reads from lxu_cache_weights (populated by previous restore)

── STASH (free cache HBM) ──────────────────────────────────
  1. flush()                        write dirty lines (from iter i-1
                                    TBE backward) → weights_uvm
  2. lxu_cache_weights.resize_(0)   free cache HBM
  3. (optionally free lxu_cache_state, lxu_state too)

── DENSE FORWARD + DENSE BACKWARD ──────────────────────────
  Cache HBM is available for activations and dense compute.

── RESTORE + PREFETCH (combined) ───────────────────────────
  4. lxu_cache_weights.resize_(original_size)
  5. reset_cache_states()           invalidate all slots (state = -1)
  6. prefetch(batch_i)              load current batch rows for TBE backward
  7. prefetch(batch_i+1)            load next batch rows for next forward

backward(batch_i)
  └─ TBE fused backward+optimizer updates lxu_cache_weights in-place
     batch_i rows are now DIRTY
     batch_i+1 rows remain CLEAN

                    Iteration i+1

forward(batch_i+1)
  └─ reads batch_i+1 rows from cache (loaded at step 7, still clean)

── STASH ─────────────────────────────────────────────
  1. flush()                        write batch_i dirty rows → weights_uvm
  2. lxu_cache_weights.resize_(0)   free cache HBM
  ...
```

### 3.2. First Iteration (Cold Start)

On iteration 0, there is no prior restore. A normal prefetch runs before the first forward:

```
prefetch(batch_0)       ← standard cache populate
forward(batch_0)
── stash ──
  flush() + free cache
── dense ──
── restore + prefetch ──
  prefetch(batch_0)     ← for backward
  prefetch(batch_1)     ← for next forward
backward(batch_0)
```

## 4\. Cache State at Each Phase

| Phase | Cache Contents | Dirty? |
| :---- | :---- | :---- |
| After forward(batch\_i) | batch\_i rows \+ batch\_i+1 rows \+ historical | batch\_i-1 rows (from prior TBE backward) |
| After flush() | same rows | all clean |
| After resize\_(0) | empty | — |
| After restore \+ prefetch | batch\_i \+ batch\_i+1 only | all clean |
| After TBE backward(batch\_i) | batch\_i \+ batch\_i+1 | batch\_i rows dirty |
| After forward(batch\_i+1) | batch\_i \+ batch\_i+1 | batch\_i rows dirty |

### 5.1. No D2H/H2D Copy of Cache Data

Unlike standard embedding stashing (which copies weights to CPU and back), this design:

* Frees cache HBM without copying to CPU
* Repopulates from weights\_uvm (authoritative after flush) via prefetch
* Avoids the cost of copying the entire cache to/from CPU

The cost is a UVM→HBM read for |batch\_i unique rows| \+ |batch\_i+1 unique rows| at restore time, which is much smaller than the full cache.

### 5.2. Cache Holds Only Active Rows

Standard EMO cache holds cache\_load\_factor \* total\_rows rows persistently. This design only holds rows for two batches. Benefits:

* No wasted HBM on historical rows that aren't needed
* Potential to allocate a smaller cache (fewer cache\_sets) since we only need to fit two batches

### 5.3. lxu\_cache\_locations Re-mapping

After reset\_cache\_states() \+ prefetch(), rows land in different cache slots than the original forward. The lxu\_cache\_locations produced by the re-prefetch in step 6 must be used by the TBE backward instead of the original locations from before forward.

Options:

* Update lxu\_cache\_locations\_list — replace the stored locations for batch\_i with new ones from step 6
* Store new locations separately — the restore step produces new locations that get consumed by backward

### 5.4. Next Batch Index Availability

Step 7 requires batch\_i+1 indices at restore time. In TorchRec's pipelined training, start\_sparse\_data\_dist(batch\_i+1) typically runs during forward, so indices should be available by restore time.

If not available (e.g., last batch), skip step 7 — the next forward will handle its own prefetch or the pipeline drains.

### 6.1. FBGEMM TBE Changes

The current TBE prefetch() is designed to run before forward. For this design:

* prefetch() API itself doesn't change — it already takes indices/offsets and populates the cache
* reset\_cache\_states() already exists and handles cache invalidation
* flush() exists as torch.ops.fbgemm.lxu\_cache\_flush

No fundamental TBE API changes needed — the orchestration changes happen at the pipeline level.

### 6.2. Pipeline-Level Changes (TorchRec)

The TrainPipelineSparseDistEmbStash pipeline needs to:

1. Identify EMO tables — tables using fused\_uvm\_caching compute kernel
2. At stash time:
   1. Call flush() on the TBE module
   2. resize\_(0) on lxu\_cache\_weights (and optionally metadata tensors)
3. At restore time:
   1. resize\_ back to original size
   2. reset\_cache\_states()
   3. Call prefetch(batch\_i\_indices) — get new lxu\_cache\_locations for backward
   4. Call prefetch(batch\_i+1\_indices) — get lxu\_cache\_locations for next forward
   5. Wire the new lxu\_cache\_locations into the backward path

### 6.3. prefetch\_pipeline Interaction

With prefetch\_pipeline=True, the TBE uses lxu\_cache\_locking\_counter to prevent eviction of in-flight rows. Since we flush \+ reset the entire cache, the locking counter state is irrelevant — we start fresh. But:

* The locking counter should also be reset at step 5
* After re-prefetch in steps 6-7, locking should work normally for the two batches

### 6.4. Memory Metadata Tensors

Beyond lxu\_cache\_weights, consider freeing:

| Tensor | Size | Worth freeing? |
| :---- | :---- | :---- |
| lxu\_cache\_weights | cache\_sets \* 32 \* max\_D \* dtype | Yes — primary HBM consumer |
| lxu\_cache\_state | cache\_sets \* 32 \* 8 bytes | Small — probably not worth the complexity |
| lxu\_state | cache\_sets \* 32 \* 8 bytes | Small — probably not worth the complexity |
| lxu\_cache\_locking\_counter | cache\_sets \* 32 \* 4 bytes | Small — probably not worth the complexity |

## 7\. HBM Savings Example (Per Rank)

Assume a single rank holds 10 GB of embedding tables in UVM, with cache\_load\_factor=0.2 (default):

```
Total UVM weights on this rank:          10 GB
EMO cache (20% of table):                10 GB * 0.2 = 2 GB  (lxu_cache_weights)
Cache metadata (state, LRU, counters):   ~50 MB               (small, ignore)

Standard EMO HBM usage:                  ~2 GB  (cache lives in HBM permanently)
```

With the unified stash design:

```
During dense forward + backward:          0 GB   (cache freed)
During restore + sparse backward:        ~2 GB   (cache re-allocated, rows re-prefetched)
```

HBM freed during dense compute: \~2 GB per rank.

This is significant — dense forward/backward is often the peak HBM consumer due to activations. Freeing 2 GB of cache HBM during that phase can be the difference between fitting a model and OOMing.

Note: The re-allocated cache is the same full size (cache\_sets \* 32 \* max\_D \* dtype), but only slots for the current \+ next batch are populated. The rest are empty (state \= \-1). A future optimization could allocate a smaller cache size to two batches, but that requires deeper FBGEMM changes.

## 8\. Risks and Open Questions

1. **Re-prefetch latency** — The restore step is potentially slower and more expensive than standard embedding stashing restore. Standard stashing does a contiguous H2D copy from pinned CPU memory (\~32 GB/s effective). Re-prefetch reads from UVM (managed memory) via lru\_cache\_populate, which does per-row scatter copies through CUDA's UVM page fault mechanism — slower than raw DMA. However, the re-prefetch can run on a separate CUDA stream and overlap with dense backward (which executes before TBE backward in the backward graph). As long as the dense backward is longer than the prefetch, the latency is fully hidden. Need to benchmark whether the overlap is sufficient for large models.
2. **Cache sizing** — In standard EMO, cache size is a careful tradeoff: bigger cache \= better hit rate but more permanent HBM usage. With this design, the tradeoff is relaxed since the cache is freed during dense compute. The only requirement is that the cache is large enough to hold 2 batches of unique rows simultaneously (batch\_i for backward \+ batch\_i+1 for next forward). In practice, this is much smaller than the default cache\_load\_factor \* total\_rows, so existing cache configurations are more than sufficient. A future optimization could reduce cache\_load\_factor (or set cache\_sets directly) to allocate a smaller cache, since we no longer need it to hold a large hot set across iterations.
3. **lxu\_cache\_locations wiring** — Ensuring backward uses of the re-prefetched locations (not the original forward locations) requires careful plumbing through the pipeline.
4. **Compatibility with prefetch\_pipeline** — Need to verify that resetting lxu\_cache\_locking\_counter and re-prefetching two batches works correctly with the locking protocol.
5. **Smaller cache allocation** — Could we allocate a smaller lxu\_cache\_weights since we only need two batches? This would require changing cache\_sets at init time based on expected batch size, which is a deeper FBGEMM change.
